src/core/thread/mle_router.cpp

```diff
@@ -4118,8 +4118,8 @@ void MleRouter::HandleAddressSolicitResponse(Coap::Message *         aMessage,
 
 bool MleRouter::IsExpectedToBecomeRouter(void) const
 {
-    return IsRouterEligible() && !IsRouterOrLeader() &&
-           (Get<RouterTable>().GetActiveRouterCount() < GetRouterUpgradeThreshold()) && !mAddressSolicitRejected;
+    return IsRouterEligible() && !IsRouterOrLeader() && !mAddressSolicitRejected &&
+           (GetRouterSelectionJitterTimeout() != 0 || mAddressSolicitPending);
```

Comment:
Could you explain more on why we need to check these conditions? @librasungirl 

More thoughts on this:
- We used `!IsRouterOrLeader()`, but do you think we should change it to `IsChild()` ?
- Should we also check `mRouterTable.GetActiveRouterCount() < mRouterUpgradeThreshold` ?

Reply:
Regarding the changes, firstly `GetRouterSelectionJitterTimeout() != 0` should be equivalent with ` Get<RouterTable>().GetActiveRouterCount() < GetRouterUpgradeThreshold()`, right?

The real addition in this PR is `mAddressSolicitPending` which represent the device is in the process of becoming router (waiting for response) in case 1) there might be CoAP retransmissions and it might take several seconds for the response depends on the communication 2) the Router Eligible device is trying to become router due to `kHaveChildIdRequest` or `kParentPartitionChange` regardless the active router number v.s. mRouterUpgradeThreshold.

Agree `IsChild()` could be applied here.

Reply:
Agree that in case 2), `IsExpectedToBecomeRouter()` should return true. 

> Regarding the changes, firstly `GetRouterSelectionJitterTimeout() != 0` should be equivalent with ` Get<RouterTable>().GetActiveRouterCount() < GetRouterUpgradeThreshold()`, right?

It's not? When the REED is sending ADDR_SOL.req, `GetRouterSelectionJitterTimeout() should return 0 but `Get<RouterTable>().GetActiveRouterCount() < GetRouterUpgradeThreshold()` should still be true. 

Overall I agree the new code is more better.

Reply:
yes, you are right. My saying is not so precise, I intended to say they are equivalent before sending ADDR_SOL.req if the REED finally would become router.
Thanks.